Symex: ignore null dereferences when targeting Java

smowton opened the pull request:
In Java (and other safe languages) we know that any path that led to dereferencing
a null pointer would have thrown or asserted beforehand (i.e. all pointer dereferences
are checked). Therefore when dereferencing a pointer with value-set {NULL, someobj}
there is no need to generate (for example)
`ptr == &someobj ? someobj.somefield : invalid_object.somefield`, but rather we can simply
produce `someobj.somefield`.

In principle this could also be extended to C programs that have `--pointer-check` enabled,
to other safe languages, and to specific pointer accesses that are analysed never-null at a
particular program point.

tautschnig commented:
So why do we have `null` in the set of objects in the first place?

smowton commented:
Flow insensitivity.

```java
A a = null;
if(some_param)
  a = new A(5);

// VS for a = { dynamic_object1, NULL }

int val = a.value;
```

Here propagating `5` to `val` is safe because the field `a.value` can only be accessed if `a` is non-null. The code is short-hand for:

```java
A a = null;
if(some_param)
  a = new A(5);

if(a == null)
  throw new NullPointerException();
int val = a.value;
```

The result is a needlessly complex `val#1 == a == &dynamic_object1 ? 5 : a$invalid_object.value` expression, when we could just have `val#1 == 5` and also constant propagation.

Generally symex could figure this out by using the guard at `int val = ...` (`a != null`) to filter the value-set. However in a safe language it seems simpler to observe that all pointer accesses are checked.

smowton commented:
Similar to the other PR, if you don't want to see `ID_java` in goto-symex, how about attaching an attribute `ID_C_is_checked` to `dereference_exprt`, which indicates a check has already been made to ensure we cannot dereference a null pointer, whether an assert, assume, exception or ordinary branch?

smowton commented:
Discussed in person with @kroening: agreed solution is for the driver program to pass a parameter to `bmct` -> `goto_symext` indicating whether it can guarantee checked dereferences-- for `jbmc` this will always be true, and for `cbmc` it will be true if `pointer-checks` is set.

smowton commented:
@tautschnig @peterschrammel @kroening @mgudemann this has been updated with a first shot at the design described above and is ready for review.

tautschnig commented:
I believe we first need to converge on the discussion in #2031. No need to get the code in that PR merged, but the semantics of `ASSERT` need to be clarified. The code currently generated is
```
        // 24 file test.java line 14 function java::test.main:(Z)V bytecode-index 10
        // Labels: pc20$
        maybe_t = (struct test *)$stack_tmp1;
        // 25 file test.java line 20 function java::test.main:(Z)V bytecode-index 13
        ASSERT !(maybe_t == null) // Throw null
        // 26 file test.java line 20 function java::test.main:(Z)V bytecode-index 13
        field_value = maybe_t->field;
```
and the proposal in this PR would only apply if the semantics of `ASSERT` is "you can't get past 'assert(false)'" (to quote @smowton).

Edit: if the front-end directly tells the back-end what the semantics of the intermediate representation is then we're violating established compiler design principles. Maybe there are good reasons for doing so, but I'd rather see those reasons first. So let's please sort out the semantics of the above snippet first.

smowton commented:
Interesting. In that case jbmc should only pass this flag (at present) when running in --cover mode (when assertions-to-assumptions means we really can't pass the assertion).

Reason for the ugly design: because the general problem "can this pointer possibly be null at this program point?" is quite hard to answer -- in principle the state guard tells us the answer, but extracting it might be difficult when e.g. trying to check whether `x->y->z` is known to be not null, which might manifest as a great big explosion of ternary conditionals over possible referees of `x` and `y`. When running against a safe language however we know this is all wasted effort, as the front-end undertakes to always check pointer derefs.

smowton commented:
@tautschnig commented at #2031; perhaps we can make assert sane rather than having to restrict the applicability of this patch.

martin-cs commented:
I guess after all of the ranting it's clear what my opinion / suggestion / preferred solution for this is...

smowton commented:
@tautschnig now the fatal-pointer-checks patch is in, are you happy for this to move forwards?

tautschnig reviewed (dismissed):
This must not be merged as is, for two reasons:
1) It is unsound: a concurrent update setting a reference to null would be missed.
2) The clarified semantics of assert/assume now enable tweaking either symex or (preferred) the value-set analysis to take into account non-nullness constraints. (Most likely it will be a combination of both: symex should update the value-set analysis upon processing an assume.) This can be done soundly, even in presence of concurrency.

smowton commented:
1. Would it be sufficient for the front-end to determine if it's submitting a concurrent program? I have no idea how the Java frontend would actually generate one, but presumably one of the former Java concurrency team knows. This is really a shortcoming of the frontend, as there is no atomic check-not-nulll-and-dereference operation -- any data race over a pointer can produce results that are in fact impossible on the JVM.
2. `assume(x != null)` filters the value-set sounds tempting, but is very low-powered, as e.g. `assume(x->y->z != null); x->y->z->w = 5` will hardly ever eliminate the null pointer, as `get_reference_set(x->y->z)` is probably not singular. Really you want a peephole alias analysis that checks that `x->y->z` must alias the `x->y->z` used some number of ASSUME instructions ago (not strictly the preceding instruction, due to `ASSUME(x != null); ASSUME(x->y != null); ASSUME(x->y->z != null); // access x->y->z`).

We could get reasonable power by adding a new data structure to symex to track expressions that are currently known not null (blanked on any assignment or control flow, added to by assume-not-null statements), but perhaps this is overpowered for the desired result?

tautschnig commented:
> any data race over a pointer can produce results that are in fact impossible on the JVM.

If that's the case, then the assert+assume+access should be placed in an atomic section.

I think there is a fair amount of C code that does `if(p != NULL) *p = 42;` and symex should be able to handle this nicely - not just in the Java case. As much as I can, I will keep pushing back against front-end <-> symex communication. We do symbolic execution on goto programs, and if there are code patterns that cause unnecessary slow-down we should develop an analysis to detect those, and not use some front-end hints. We already do pass the `state.guard` to the value-set analysis, but I don't think it makes as much use of it. I think we should be able to make at least the simple example described in this PR resolve via this route. This requires studying what information `state.guard` actually has, and what the value-set analysis does about it.

smowton commented:
I think I can nail something together that does the trick for the most basic cases of null pointer checks without that being either (a) unduly costly or (b) unduly complex. However what do you want to do about the integer address case? That was already worked around with a simple `if(language_mode == ID_java)`, and there aren't going to be source-level assertions about that -- it's just a static universal property of Java, and of GOTO generated from Java, much like all derefs being checked.

tautschnig commented:
> However what do you want to do about the integer address case? 

I'm actually surprised this case ever arises in code originating from Java? In the commit message back then you said that you weren't able to reproduce it anymore, so I'm wondering whether 1b2e321 should actually be reverted?

smowton commented:
I'll have a go tomorrow. My guess is a may-alias between a primitive array and a reference array, but I'll see if I can make anything fly...

smowton commented:
@tautschnig here's a draft of what I'd do if front-end promises aren't enough. It's a really simple analysis that can spot ASSUME and very simple conditional-GOTO null checks -- I wanted to avoid either another analysis integrated with symex, or a potentially expensive fixpoint analysis.

https://github.com/diffblue/cbmc/compare/develop...smowton:smowton/feature/symex-exclude-null-pointers-take-two?expand=1

Could you let me know if this loose direction is acceptable before I go any further?

tautschnig commented:
Hmm ... "local safe pointers" - are you sure the `local_may_aliast` can't be taught as much? Though really at this point I'm more in need of data points than in taking a look at some implementation. The data we'll need in order to have a meaningful discussion is, at the bare minimum: for your above example, at the point of dereferencing, 1) the contents of the value set and 2) the state guard? Such data will permit answering the following basic questions:
1) Could the value set make that decision if reasoning more precisely about guards?
2) Does symex have sufficient information to make a decision or channel that information to the value set to enable it to make a decision?

smowton commented:
@tautschnig 

Example program:
```java
public class test {

  public int field;
  public test ptr;

  public static void main(boolean unknown1, boolean unknown2, boolean unknown3, int someint) {

    test root = null;
    if(unknown1) {
      root = new test();
      if(unknown2) {
        root.ptr = new test();
        if(unknown3) {
          root.ptr.ptr = new test();
          root.ptr.ptr.field = someint;
        }
      }
    }

    int local = root.ptr.ptr.field;

    assert false;

  }

}
```

The value set is roughly (renaming dynamic objects for clarity):
```
root = { null, &ind1 }
ind1.ptr = { null, &ind2 }
ind2.ptr = { null, &ind3 }
ind3.ptr = { null }
```

The constraint formed by the first not-null assumption is readily understood:
`!(root!0@1#4 == null)`
But for subsequent nested derefs it becomes trickier:
`!(root!0@1#4 == &dynamic_object1 ? dynamic_object1#5.ptr == null : root$object#0.ptr == null)`
and then:
`!((root!0@1#4 == &dynamic_object1 ? dynamic_object1#5.ptr == &dynamic_object2 : root$object#0.ptr == &dynamic_object2) ? dynamic_object2#6.ptr == null : invalid_object1#0.ptr == null)`

The state guard is similarly kept in this expanded form. This is more easily seen with a conditional than an assumption:

```java
  public static void main(boolean unk1, boolean unk2, boolean unk3) {

    x x1 = new x();
    if(unk1)
      x1.field = new x();

    x x2 = new x();
    if(unk2)
      x2.field = new x();

    x root = unk3 ? x1 : x2;

    if(root.field != null) {

      assert false;

    }

  }
```
yields a guard: `\guard#5 == !(root!0@1#2 == &dynamic_object3 ? dynamic_object3#4.field == null : dynamic_object1#4.field == null)`

The value-set is definitely unsuitable for tracking not-null assumptions, because in the previous example where would I record it? I can't say "dynamic_object3#4.field != null" or "dynamic_object1#4.field != null", as neither one of those is true in general -- what matters is the condition (`root!0@1#2 == &dynamic_object3`) is the same as the one on the assumption, or in other words, the pointer assumed not-null must alias the one being dereferenced, with no possible modifications between the two instructions.

The assumption condition on the other hand is in principle suitable, as it contains the right information. However, (1) assumptions are not recorded into the state guard for non-threaded programs, but rather are simply emitted as guarded constraints (see `goto_symext::symex_assume`), so recovering that information at the dereference site requires walking back to search for the relevant assume, either in the GOTO program or the equation produced so far, and (2) the assumption has been exapanded (put through symex's clean_expr including dereferencing), meaning either we must hope that the simplifier gives us the same expansion / simplification in the deref and the assumption, or else we must try to check whether the assumption implies the deref'd pointer cannot be null, which sounds difficult enough that it's really a job for the solver backend, not symex.

Long story short, the value-set is the wrong structure for the task, and recovering the desired information from the expanded form of the assumption is possible but I think will be brittle and quite expensive. It is much easier, by contrast, to extract the desired information from the GOTO program, as we have a simple must-alias relationship between the assumed pointer and the dereferenced one with only no-mod-ref instructions in between (other assumes in Java; control flow in the simple C `x ? *x : 0` case).

So, if we are to recognise a safe-deref operation at the GOTO level, we have a couple of choices:
1. Compute in-line with symex, or write a standalone GOTO analysis? The latter is nicer for modularity, but the former has often been done so far. I don't mind too much which we choose.
2. Are any existing GOTO analyses suitable? 

`local_bitvector` has a go at diagnosing pointers that aren't null, but it gets its information from assignments and indexes it by symbol identifier, both of which are unsuitable for this use case -- it is intended to avoid a check on `int x; int *y = &x; *y = 20;`, but can't spot `assume(complicated_expr !=null); complicated_expr->y = 5;`, as its domain can't record `complicated_expr` in any form.

`local_may_alias` also exists, but it appears to be unsound.

What we need here is a very simple, cheap must-alias analysis (which would also be useful to goto-check incidentally, as a complement to local-bitvector), which would determine when there are two accesses in sequence that must alias one another.

One final caveat: I note that symex would also be "trusting" an external module here. I might also write a safe-pointer "analysis" that just checks if ns.lookup(instr->function).mode == ID_java... but I get that this sort of analysis would also spot other guarded accesses :)

smowton commented:
@tautschnig thinking on this further, the frustration of trying to write an analysis for this is that we're recovering information in symex that we knew in the frontend but then needlessly threw away. Therefore, suggested alternative approaches:

1. Checked-dereference primitive -- a C analysis pass could try to promote derefs to checked derefs when it can prove that's sound; Java makes such derefs out of the box; the semantics of checked_deref(x) is the same as atomic { assume(x != null); *x }; symex natively supports the operation. Probably implemented as an attribute on dereference_exprt.
2. Same as above but avoids inflating every deref expr: the frontend can run symex in checked-derefs mode, which changes the semantics of all deref expressions. Less versbose, but I don't like the idea of the same expression having several meanings like this.
3. Java frontend could always do derefs via a temporary local (e.g. `derefd_ptr = <complicated expression>; assume(derefd_ptr != null); <use *derefd_ptr>`), then teach symex to filter the value-set on a not-null assumption. This is slightly more general, as that kind of assume could in theory be given by a C program, but has the disadvantage of significantly inflating Java-dervied GOTO programs.

My order of preference at the moment:

1. Checked deref primitive (option #1 above)
2. Very simple checked deref analysis (as prototyped last night)
3. Deref via temporaries (option #3 above) + value-set filtering
4. Blunt assertion by the frontend (solution suggested in this PR)
5. Extracting not-null constraints from guards in symex

smowton commented:
@tautschnig ping; a briefer version of the above: what do you think to a checked-deref primitive?

peterschrammel commented:
> assumptions are not recorded into the state guard for non-threaded programs

Would changing this help?

smowton commented:
@peterschrammel I don't think it would help much --since the guards are recorded in simplified-and-dereferenced form, re-establishing the correspondence between the assume and the deref will still be harder than doing it from the raw GOTO, which in turn it harder than getting that information from a primitive checked-deref operation, hence my order of preferences above

smowton commented:
@tautschnig @peterschrammel conversation with @kroening today concluded that a peephole GOTO analysis, structured similarly to `dirtyt`, would be the least bad option -- it stands a chance of spotting user as well as generated checks, it's pretty cheap, and it's modular and usable away from symex -- for example, goto-check could use it to avoid redundant checks similarly to its existing use of `local_bitvectort`. Any objections to going forward with that route?

tautschnig commented:
With apologies for the late response: as much as I'd prefer a smart(TM) analysis that re-instantiates all the information: as a variation on the proposed check-deref annotation, maybe introduce a new `checked_deref_exprt : public exprt` that does the same?

Now as I am posting this concurrently with @smowton reporting on his conversation: I do strongly support doing an analysis pass. If you can throw in all your compiler knowledge and do this: would be brilliant!

smowton commented:
@tautschnig @kroening this is now updated with the new strategy agreed above and is ready to review.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -0,0 +1,239 @@
+
+#include "local_safe_pointers.h"
+
+#include <langapi/language_util.h>
```
Use `format` (and `format_expr.h`) instead.

tautschnig commented on regression/cbmc/symex_should_exclude_null_pointers/main.c:
```diff
@@ -0,0 +1,74 @@
+
+  // Shouldn't work due to suspicion *ptr6 might alter ptr7:
+  int *ptr7 = maybe_null;
+    __CPROVER_assume(ptr7 != 0);
```
Nit picking: spurious indent (occurs multiple times).

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -462,6 +464,8 @@ class goto_symext
 
   path_storaget &path_storage;
+
+  std::map<irep_idt, local_safe_pointerst> safe_pointers;
```
Nit pick: `std::unordered_map`?

tautschnig commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -236,6 +236,22 @@ void goto_symext::dereference_rec(
+
+    if(state.threads.size() == 1)
+    {
+      irep_idt expr_function = state.source.pc->function;
```
`const irep_idt &`

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -130,10 +130,17 @@ void goto_symext::initialize_entry_point(
   state.top().calling_location.pc=state.top().end_of_function;
   state.symex_target=&target;
 
+  const goto_functiont &entry_point_function = get_goto_function(pc->function);
```
Shouldn't this be placed after the `INVARIANT` below?

tautschnig commented on src/analyses/local_safe_pointers.h:
```diff
@@ -0,0 +1,57 @@
+      return false;
+    exprt tocheck = expr;
+    while(tocheck.id() == ID_typecast)
+      tocheck = tocheck.op0();
```
We typically use pointers in such cases to avoid repeated construction of short-lived objects.

tautschnig commented on src/analyses/local_safe_pointers.h:
```diff
@@ -0,0 +1,57 @@
+    exprt tocheck = expr;
+    while(tocheck.id() == ID_typecast)
+      tocheck = tocheck.op0();
+    return findit->second.count(tocheck);
```
Please use `findit->second.find(tocheck) != findit->second.end()` (or, if you insist, add `!= 0` to your code -- Visual Studio issues warnings for your code.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -0,0 +1,239 @@
+       op1 == null_pointer_exprt(to_pointer_type(op1.type())))
+    {
+      while(op0.id() == ID_typecast)
+        op0 = op0.op0();
```
Use pointers instead of short-lived objects.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -0,0 +1,239 @@
+#include <util/expr_iterator.h>
+
+/// If `expr` is of the form `x != nullptr`, return x. Otherwise return nil.
+static exprt get_null_checked_expr(const exprt &expr)
```
You really want to use pointers in the below code and have `const exprt &` as return type.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -0,0 +1,239 @@
+    }
+  }
+
+  return nil_exprt();
```
To comply with my recommendation of returning a `const exprt &` do `return static_cast<const exprt&>(get_nil_irep());`

smowton replied:
I've just used a pointer as a "poor man's optional," but otherwise done as you suggest.

tautschnig reviewed (commented):
This is a nice piece of work, and the fact that you've added a lot of documentation makes it even better :-)

Before approving it, though, I'd like to know what the performance impact is? Have you done any measurements how much longer a symex run takes? I do acknowledge that this isn't unambiguous: the added pointer-dereferencing precision should make up for some of the time cost of running the analysis for each of the functions encountered.

smowton committed:
Add local-safe-pointers analysis

This approaches the safe-pointers problem from the opposite direction compared to local-
bitvector-analysis -- local-bitvector tries to establish a pointer is safe because it
definitely points to known object, whereas local-safe-pointers tries to establish a
particular usage is certainly safe because it is trivially dominated by a not-null test,
which at present means a conditional GOTO or an ASSUME.

smowton committed:
Use local-safe-pointers analysis to improve Symex pointer resolution

This uses local_safe_pointerst to determine when symex dereferences a pointer
that cannot be null. When it does the null result is excluded from the possible
values, and therefore a $invalid_object reference may be excluded from the result
of dereferencing such a pointer. This can improve constant propagation.

smowton commented:
@tautschnig all comments applied. Benchmarking coming up...

smowton commented:
Benchmark: I generated a synthetic benchmark using this script:
```python
#!/usr/bin/env python

import random
import sys

if len(sys.argv) != 2 or sys.argv[1] not in ("good", "bad"):
    print >>sys.stderr, "Usage: gen.py (good | bad)"
    sys.exit(1)

spoil_assumption = sys.argv[1] == "bad"

print "int main(int argc, char **argv) {"

for i in range(100):
    print "  int local%d = %d;" % (i, i)

vars = ["&local%d" % i for i in range(100)]

choices = 0
def tchoice(l):
    n = len(l)
    global choices
    out = ""
    while len(l) > 1:
        out += "(argc == %d ? %s : " % (choices, l[-1])
        choices += 1
        l.pop()
    if len(l) == 1:
        out += "(%s" % l[0]
    out += ")" * n
    return out

print "  "

for i in range(10000):
    naliases = random.randint(1, 5)
    aliases = random.sample(vars, naliases) + ["0"]
    print "  int *ptr%d = %s;" % (i, tchoice(aliases))
    print "  __CPROVER_assume(ptr%d != 0);" % i
    if spoil_assumption:
        print "  int incidental%d = 0;" % i
    print "  int deref%d = *ptr%d;" % (i, i)

print "  "
print "  assert(0);"

print "}"
```

In the "good" case it generates something of  the form
```c
int local1 = 1, local2 = 2, ...;
int *ptr1 = nondet-choice-of(&local1, &local2, ...., null);
__CPROVER_assume(ptr1 != 0);
int deref1 = *ptr1;
```

and in the bad case it adds an extra `int incidental1 = 0;` right before the deref, to confound the analysis and cause it to make no improvement.

Results (Old and New are cbmc before and after this patch; Good and Bad are the scenarios above where the optimisation works and doesn't work respectively):
Time (s):

| |Old|New|
|---|---|---|
|Good|64|62|
|Bad|76|81|

Size of VCCs (printed, million characters):

|     |Old|New|
|---|---|---|
|Good|3.7|3.2|
|Bad|4.1|4.1|

In all cases I used the debug build (for conveinence) and `--show-vcc` because the solver didn't terminate in a reasonable timescale. It seems reasonable to suppose that with 3.2M rather than 3.7M chars of VCCs in the Good case it would have recouped some time. However even within the symex-and-print-vccs process we can see a small (6.5%) time cost in the bad case, where we gather lots of information about possible safe pointers but never get to use that information profitably, and a small benefit (3%) in the good case where our information always makes us more certain about a pointer, and so we emit a smaller formula. That benefit would become larger if the result of the deref was subsequently used in expressions that could be constant-propagated away.

tautschnig commented:
@smowton Thanks a lot for the detailed benchmarking report! So it seems that in case of C it doesn't make a whole lot of a difference (some improvement, some cost). Hence I'm now curious about results on the benchmarks/problems you had originally looked at that actually lead to you starting this PR?

smowton commented:
To give a simple example, I added a new "best" case to the script, which always uses alias sets of size 1 + null, and includes some junk code only executed when we're uncertain -- for example, `if(deref1 != local30) { for(int j = 0; j < 10; ++j) { } }`

So the situation looks like
```c
int x = 0;
int *ptr = nondet(int) ? &x : null);
assume(ptr != null);
int x2 = *ptr;
if(x != x2) { // something expensive }
```

I had to lower the number of ops to 1000 (from 10000) for this version to complete in a reasonable timeframe.

Predictably this takes the old CBMC 42s / 1.4Mc of VCCs, and the new one 6s / 0.16Mc of VCCs.

So very large gains are possible when improved constant prop results in certainty about a control-flow choice. The most common form of this in Java is a virtual function call (e.g. `A a = nondet(int) ? null : new A(); a.f();`). That code can only call the `f()` method implemented or inherited by `A`, but in the case where symex thinks there may be a null pointer deref it includes `a$object.@class_identifier` as a possible key used to dispatch the call, and being uncertain about that value also considers `B.f()`, `C.f()`, ... as possible callees. It then generates VCCs for those in-fact-unreachable callees, resulting in potentially enormous cost, depending on how nasty `B.f`, `C.f` et al are.

By itself I suspect this won't deliver a massive improvement on practical Java software, as without https://github.com/diffblue/cbmc/pull/2122 we become uncertain about `@class_identifier` because it is a field of a structure rendered non-constant for other reasons. However, combined with other improvements like that we can increase our chance of resolving a virtual function target (or perhaps a function pointer target?) at symex time, and therefore exclude great gluts of code from the target formula.

tautschnig commented:
Ok, thanks, so what is your preferred way to proceed? Should this wait until we have all the bits together to demonstrate big gains, or should we try to merge this soon/now? (I'd actually prefer to hold off on such a change until after 5.9 is released, which should be happening really soon -> @kroening). My suggestion: discussion (offline) between @smowton and @kroening.

smowton commented:
Personally I'd like to get this merged as big merges are very difficult. Re: #2122 we need to decide-- is your field-sensitive constant prop stuff going to get in any time soon, and will it be good enough to always resolve a clsid test? If it's too difficult and too performance sensitive to get it in, can we special-case it in a way that's less ugly than the rejected #2122 (which used magic knowledge that `@class_identifier` is just another word for dynamic type)? Can we stop using a field to store the RTTI at all, and instead introduce a primitive RTTI operator that symex can cleanly interpret?

smowton commented:
Discussed with @kroening just now: he is to confer with @tautschnig re: when this goes in; I will follow up with a separate PR to re-use the `guard` already passed down the `symex_dereference` recursion to exclude null pointers from expressions like `ptr ? *ptr : 0`, rather than producing silly expressions like `ptr ? (ptr == &local ? 1 : ptr$object) : 0`

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: aa1b24958637e09840c781bb16cb904ebeee2c21).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76979426

peterschrammel commented on regression/goto-instrument/safe-dereferences/main.c:
```diff
@@ -0,0 +1,71 @@
+
+  // Shouldn't work due to suspicion *ptr6 might alter ptr7:
+  int *ptr7 = &x;
+    __CPROVER_assume(ptr7 != 0);
```
weird indentation (several occurrences)

peterschrammel commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -0,0 +1,238 @@
+#include <util/format_expr.h>
+
+/// If `expr` is of the form `x != nullptr`, return x. Otherwise return null
+static const exprt *get_null_checked_expr(const exprt &expr)
```
use `optional` here as well?

smowton replied:
Michael argued for a pointer earlier, so I'm going to stick with that rather than go back and forth :)

peterschrammel replied:
Alright

peterschrammel commented on src/analyses/local_safe_pointers.h:
```diff
@@ -0,0 +1,57 @@
+/// In the interests of a very cheap analysis we only search for very local
+/// guards -- at the moment only `if(x != null) { *x }`
+/// and `assume(x != null); *x` are handled. Control-flow convergence and
+/// possibly-alising operations are handled pessimistically.
```
aliasing

peterschrammel commented on src/analyses/local_safe_pointers.h:
```diff
@@ -0,0 +1,57 @@
+ public:
+  void operator()(const goto_programt &goto_program);
+
+  bool is_non_null_at_program_point(
```
Is this sizeable method intentionally defined in the header file?

peterschrammel commented on jbmc/regression/jbmc/symex_should_exclude_null_pointers/test.desc:
```diff
@@ -0,0 +1,13 @@
+--
+^warning: ignoring
+--
+The test should fail, as it might dereference a null pointer, but as Java is a safe language we should
```
Nitpick: I guess JBMC should report FAILED (not the 'test should fail').

peterschrammel reviewed (approved):
LGTM, some nitpicks

allredj commented:
@smowton do you plan on doing a test-gen PR for that?

smowton commented:
ok

smowton commented:
https://github.com/diffblue/test-gen/pull/1973

smowton commented:
Applied all of Peter's comments apart from one response

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ab421372c0ba1c794c920d346e4e79d2dcfc03bd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77037324

allredj commented:
Bump is failing some regression tests.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9fd3434c96b68042824d99bab5d46a7cc82a6381).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77151625
